File: 算法/贪心算法/num55.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int len = nums.size();
        if (len==0) return false;
        if (len==1) return true;
        bool dp[len] = {false};
        dp[len-1] = true;
        for(int i=len-2; i>=0; i--){
            int futhestJump = nums[i]+ i;
            if (futhestJump > len-1) futhestJump = len -1;
            // check whether can jump to terminal from i
            for (int j=i+1; j<= futhestJump; j++){
                if (dp[j] == true){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[0];
    }
};

File: 算法/动态规划/num55.py
class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        MaxJump = 0
        for i in range(len(nums)):
            if i > MaxJump:
                return False
            MaxJump = max(MaxJump, nums[i]+i)
        return True